edgedb.go

```diff
@@ -74,36 +81,98 @@ func connectOne(ctx context.Context, cfg *connConfig, conn *baseConn) error {
 		err error
 	)
 
+	conn.writer = buff.NewWriter()
+
 	for _, addr := range cfg.addrs { // nolint:gocritic
-		// todo do error values need to be checked?
 		conn.conn, err = d.DialContext(ctx, addr.network, addr.address)
 		if err != nil {
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err)
+			continue
+		}
+
+		toBeDeserialized := make(chan *soc.Data, 2)
+		r := buff.NewReader(toBeDeserialized)
+		go soc.Read(conn.conn, soc.NewMemPool(4, 256*1024), toBeDeserialized)
+
+		err = conn.setDeadline(ctx)
+		if err != nil {
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
-		err = conn.connect(ctx, cfg)
+		err = conn.connect(r, cfg)
 		if err != nil {
-			_ = conn.conn.Close()
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
+		conn.acquireReaderSignal = make(chan struct{}, 1)
+		conn.readerChan = make(chan *buff.Reader, 1)
+		conn.releaseReader(r)
 		return nil
 	}
 
 	conn.conn = nil
 	return err
 }
 
+func (c *baseConn) setDeadline(ctx context.Context) error {
```

Comment:
Hm, what's this function for?

Reply:
Ah, that's for client-side timeouts, right?

Reply:
correct :heavy_check_mark: